opcua/server/internal_server.py

```diff
@@ -269,7 +281,46 @@ def set_attribute_value(self, nodeid, datavalue, attr=ua.AttributeIds.Value):
         """
         self.aspace.set_attribute_value(nodeid, ua.AttributeIds.Value, datavalue)
 
+    def set_user_manager(self, user_manager):
+        """
+        set up a function which that will check for authorize users. Input function takes username
+        and password as paramters and returns True of user is allowed access, False otherwise.
+        """
+        self.user_manager = user_manager
 
+    def check_user_token(self, isession, token):
+        """
+        unpack the username and password for the benefit of the user defined user manager
+        """
+        userName = token.UserName
+        passwd = token.Password
+        # decrypt password is we can
+        if str(token.EncryptionAlgorithm) != "None":
+            if use_crypto == False:
+                return False;
+            try:
+                if token.EncryptionAlgorithm == "http://www.w3.org/2001/04/xmlenc#rsa-1_5":
+                    raw_pw = uacrypto.decrypt_rsa15(self.private_key, passwd)
+                elif token.EncryptionAlgorithm == "http://www.w3.org/2001/04/xmlenc#rsa-oaep":
+                    raw_pw = uacrypto.decrypt_rsa_oaep(self.private_key, passwd)
+                else:
+                    self.logger.warning("Unknown password encoding '{0}'".format(token.EncryptionAlgorithm))
+                    return False
+                length = unpack_from('<I', raw_pw)[0] - len(isession.nonce)
+                passwd = raw_pw[4:4 + length]
+                passwd = passwd.decode('utf-8')
+            except Exception as exp:
+                self.logger.warning("Unable to decrypt password")
+                return False
+
+        # call user_manager if one is defined
+        if self.user_manager is not None:
+            return self.user_manager(self, isession, userName, passwd)
+
+        # If nothing function defined, do the default thing
+        if self.allow_remote_admin and userName in ("admin", "Admin"):
```

Comment:
Why did you remove the default user manager function?I thought it was a good idea. It shows people how they can make they own

Reply:
I see. I can add it back in.... where should this function live?